gcloud/monitoring/_dataframe.py

```diff
@@ -0,0 +1,114 @@
+                     label=None, labels=None):  # pragma: NO COVER
+    """Build a ``pandas`` dataframe out of time series.
+
+    :type time_series_iterable:
+        iterable over :class:`~gcloud.monitoring.timeseries.TimeSeries`
+    :param time_series_iterable:
+        An iterable (e.g., a query object) yielding time series.
+
+    :type label: string or None
+    :param label:
+        The label name to use for the dataframe header. This can be the name
+        of a resource label or metric label (e.g., ``"instance_name"``), or
+        the string ``"resource_type"``.
+
+    :type labels: list of strings, or None
+    :param labels:
+        A list or tuple of label names to use for the dataframe header.
+        If more than one label name is provided, the resulting dataframe
+        will have a multi-level column header.
+
+        Specifying neither ``label`` or ``labels`` results in a dataframe
+        with a multi-level column header including the resource type and
+        all available resource and metric labels.
+
+        Specifying both ``label`` and ``labels`` is an error.
+
+    :rtype: :class:`pandas.DataFrame`
+    :returns: A dataframe where each column represents one time series.
+    """
+    import pandas   # pylint: disable=import-error
+
+    if label is not None and labels is not None:
+        raise ValueError('Cannot specify both "label" and "labels".')
+    elif not (labels or labels is None):
+        raise ValueError('"labels" must be non-empty or None.')
+
+    columns = []
+    headers = []
+    for time_series in time_series_iterable:
+        pandas_series = pandas.Series(
+            data=[p.value for p in time_series.points],
+            index=[p.end_time for p in time_series.points],
+        )
+        columns.append(pandas_series)
+        headers.append(time_series.header())
+
+    # Implement a smart default of using all available labels.
+    if label is None and labels is None:
+        resource_labels = set(itertools.chain.from_iterable(
+            header.resource.labels for header in headers))
+        metric_labels = set(itertools.chain.from_iterable(
+            header.metric.labels for header in headers))
+        labels = (['resource_type'] +
+                  _sorted_resource_labels(resource_labels) +
+                  sorted(metric_labels))
+
+    # Assemble the columns into a DataFrame.
+    dataframe = pandas.DataFrame(columns).T
+
+    # Convert the timestamp strings into a DatetimeIndex.
+    dataframe.index = pandas.to_datetime(dataframe.index)
+
+    # Build a column Index or MultiIndex from the label values. Do not
+    # include level names in the column header if the user requested a
+    # single-level header by specifying "label".
+    level_names = labels or None
+    label_keys = labels or [label]
+    dataframe.columns = pandas.MultiIndex.from_arrays(
+        [[header.labels.get(key, '') for header in headers]
+         for key in label_keys],
```

Comment:
In general, doubly-nested list comprehensions are frowned upon because they are hard to parse. Also, this seems to break the preference for iterators over manually constructing lists.